File: hosts/macos/Sources/NatuiHost/AppBundle.swift
import CryptoKit
import Foundation

struct PackagedApp {
    let id: String
    let name: String
    let version: String
    let sourceName: String
    let source: String
}

private struct AppManifest: Decodable {
    let schemaVersion: Int
    let id: String
    let name: String
    let version: String
    let buildNumber: String
    let entry: String
    let entrySha256: String
    let protocolVersion: Int
    let minHostApi: Int
    let platform: String
    let architecture: String
}

enum AppBundleError: LocalizedError {
    case invalid(String)

    var errorDescription: String? {
        switch self {
        case .invalid(let message): message
        }
    }
}

/// Resolves and validates Contents/Resources/NatUI for a packaged .app.
/// Explicit --bundle launches are available only to the bare development host.
enum AppBundleLoader {
    private static let schemaVersion = 1
    private static let protocolVersion = 1
    private static let hostApiVersion = 2

    /// Nil means the process is a bare development host, not an .app bundle.
    static func loadIfPackaged() -> Result<PackagedApp, Error>? {
        guard Bundle.main.bundleURL.pathExtension == "app" else { return nil }

        do {
            guard let resources = Bundle.main.resourceURL else {
                throw AppBundleError.invalid("The application has no Resources directory.")
            }
            let appDirectory = resources.appendingPathComponent("NatUI", isDirectory: true)
            let manifestURL = appDirectory.appendingPathComponent("manifest.json")
            let data: Data
            do {
                data = try Data(contentsOf: manifestURL)
            } catch {
                throw AppBundleError.invalid(
                    "Cannot read app manifest at \(manifestURL.path): \(error.localizedDescription)"
                )
            }

            let manifest: AppManifest
            do {
                manifest = try JSONDecoder().decode(AppManifest.self, from: data)
            } catch {
                throw AppBundleError.invalid("Invalid app manifest: \(error.localizedDescription)")
            }

            guard manifest.schemaVersion == schemaVersion else {
                throw AppBundleError.invalid(
                    "Unsupported app bundle schema \(manifest.schemaVersion); "
                    + "this host supports schema \(schemaVersion)."
                )
            }
            guard manifest.protocolVersion == protocolVersion else {
                throw AppBundleError.invalid(
                    "App requires protocol \(manifest.protocolVersion); "
                    + "this host implements protocol \(protocolVersion)."
                )
            }
            guard manifest.minHostApi > 0, manifest.minHostApi <= hostApiVersion else {
                throw AppBundleError.invalid(
                    "App requires host API \(manifest.minHostApi); "
                    + "this host implements API \(hostApiVersion)."
                )
            }
            guard manifest.platform == "macos" else {
                throw AppBundleError.invalid(
                    "App bundle targets \(manifest.platform), not macos."
                )
            }
            guard manifest.architecture == currentArchitecture else {
                throw AppBundleError.invalid(
                    "App bundle targets \(manifest.architecture); "
                    + "this host is \(currentArchitecture)."
                )
            }
            guard !manifest.id.isEmpty, !manifest.name.isEmpty, !manifest.version.isEmpty else {
                throw AppBundleError.invalid("App manifest identity fields cannot be empty.")
            }
            guard !manifest.entry.isEmpty, !manifest.entry.hasPrefix("/") else {
                throw AppBundleError.invalid("App entry must be a non-empty relative path.")
            }

            guard let bundleIdentifier = Bundle.main.bundleIdentifier,
                  !bundleIdentifier.isEmpty else {
                throw AppBundleError.invalid("Info.plist has no application identifier.")
            }
            guard manifest.id == bundleIdentifier else {
                throw AppBundleError.invalid(
                    "App manifest identifier \(manifest.id) does not match "
                    + "Info.plist identifier \(bundleIdentifier)."
                )
            }
            guard let bundleVersion = Bundle.main.object(
                forInfoDictionaryKey: "CFBundleShortVersionString"
            ) as? String, !bundleVersion.isEmpty else {
                throw AppBundleError.invalid("Info.plist has no application version.")
            }
            guard manifest.version == bundleVersion else {
                throw AppBundleError.invalid(
                    "App manifest version \(manifest.version) does not match "
                    + "Info.plist version \(bundleVersion)."
                )
            }
            guard let bundleBuild = Bundle.main.object(
                forInfoDictionaryKey: "CFBundleVersion"
            ) as? String, !bundleBuild.isEmpty else {
                throw AppBundleError.invalid("Info.plist has no application build number.")
            }
            guard manifest.buildNumber == bundleBuild else {
                throw AppBundleError.invalid(
                    "App manifest build \(manifest.buildNumber) does not match "
                    + "Info.plist build \(bundleBuild)."
                )
            }

            let root = appDirectory.resolvingSymlinksInPath().standardizedFileURL
            let entry = root
                .appendingPathComponent(manifest.entry)
                .resolvingSymlinksInPath()
                .standardizedFileURL
            let rootPath = root.path.hasSuffix("/") ? root.path : root.path + "/"
            guard entry.path.hasPrefix(rootPath) else {
                throw AppBundleError.invalid("App entry leaves the NatUI resource directory.")
            }

            let script: Data
            do {
                script = try Data(contentsOf: entry)
            } catch {
                throw AppBundleError.invalid(
                    "Cannot read app entry at \(entry.path): \(error.localizedDescription)"
                )
            }
            let digest = SHA256.hash(data: script).map { String(format: "%02x", $0) }.joined()
            guard digest == manifest.entrySha256.lowercased() else {
                throw AppBundleError.invalid(
                    "App entry integrity check failed for \(entry.path)."
                )
            }
            guard let source = String(data: script, encoding: .utf8) else {
                throw AppBundleError.invalid("App entry \(entry.path) is not valid UTF-8.")
            }

            return .success(PackagedApp(
                id: manifest.id,
                name: manifest.name,
                version: manifest.version,
                sourceName: entry.path,
                source: source
            ))
        } catch {
            return .failure(error)
        }
    }

    private static var currentArchitecture: String {
        #if arch(arm64)
        return "arm64"
        #elseif arch(x86_64)
        return "x64"
        #else
        return "unsupported"
        #endif
    }
}

File: hosts/windows/NatuiHost/AppBundle.cs
using System.Runtime.InteropServices;
using System.Security.Cryptography;
using System.Reflection;
using System.Text;
using System.Text.Json.Nodes;

namespace NatuiHost;

internal sealed record PackagedApp(
    string Id,
    string Name,
    string Version,
    string SourceName,
    string Source);

/// <summary>
/// Resolves and validates the packaged application stored beside the
/// self-contained host. Development --bundle launches are available only
/// when the host has no embedded application.
/// </summary>
internal static class AppBundle
{
    private const int SchemaVersion = 1;
    private const int ProtocolVersion = 1;
    private const int HostApiVersion = 2;

    private const string ManifestResource = "NatuiHost.App.manifest.json";
    private const string ScriptResource = "NatuiHost.App.main.js";

    private static Assembly HostAssembly => typeof(AppBundle).Assembly;

    public static bool IsEmbedded =>
        HostAssembly.GetManifestResourceInfo(ManifestResource) is not null;

    public static bool TryLoad(out PackagedApp? app, out string? error)
    {
        app = null;
        error = null;
        JsonObject manifest;
        try
        {
            using var stream = HostAssembly.GetManifestResourceStream(ManifestResource)
                ?? throw new InvalidDataException("embedded manifest resource is missing");
            manifest = JsonNode.Parse(stream) as JsonObject
                ?? throw new InvalidDataException("manifest root is not an object");
        }
        catch (Exception ex)
        {
            error = $"Cannot read the embedded app manifest: {ex.Message}";
            return false;
        }

        if (Int(manifest, "schemaVersion") != SchemaVersion)
        {
            error = $"Unsupported app bundle schema {Value(manifest, "schemaVersion")}; "
                + $"this host supports schema {SchemaVersion}.";
            return false;
        }
        if (Int(manifest, "protocolVersion") != ProtocolVersion)
        {
            error = $"App requires protocol {Value(manifest, "protocolVersion")}; "
                + $"this host implements protocol {ProtocolVersion}.";
            return false;
        }
        if (Int(manifest, "minHostApi") is not { } minimumHostApi
            || minimumHostApi < 1
            || minimumHostApi > HostApiVersion)
        {
            error = $"App requires host API {Value(manifest, "minHostApi")}; "
                + $"this host implements API {HostApiVersion}.";
            return false;
        }
        if (Str(manifest, "platform") != "windows")
        {
            error = $"App bundle targets {Value(manifest, "platform")}, not windows.";
            return false;
        }

        var expectedArchitecture = RuntimeInformation.ProcessArchitecture switch
        {
            Architecture.X64 => "x64",
            Architecture.Arm64 => "arm64",
            _ => null,
        };
        if (expectedArchitecture is null
            || Str(manifest, "architecture") != expectedArchitecture)
        {
            error = $"App bundle targets {Value(manifest, "architecture")}; "
                + $"this host is {expectedArchitecture ?? RuntimeInformation.ProcessArchitecture.ToString()}.";
            return false;
        }

        var id = Str(manifest, "id");
        var name = Str(manifest, "name");
        var version = Str(manifest, "version");
        var entry = Str(manifest, "entry");
        var expectedHash = Str(manifest, "entrySha256");
        if (string.IsNullOrWhiteSpace(id)
            || string.IsNullOrWhiteSpace(name)
            || string.IsNullOrWhiteSpace(version)
            || string.IsNullOrWhiteSpace(entry)
            || expectedHash is null
            || expectedHash.Length != 64)
        {
            error = "App manifest is missing identity, entry, or integrity fields.";
            return false;
        }

        if (entry != "main.js")
        {
            error = $"Invalid app entry {entry}; Windows single-file apps require main.js.";
            return false;
        }

        string source;
        byte[] sourceBytes;
        try
        {
            using var stream = HostAssembly.GetManifestResourceStream(ScriptResource)
                ?? throw new InvalidDataException("embedded main.js resource is missing");
            using var memory = new MemoryStream();
            stream.CopyTo(memory);
            sourceBytes = memory.ToArray();
            source = new UTF8Encoding(
                encoderShouldEmitUTF8Identifier: false,
                throwOnInvalidBytes: true).GetString(sourceBytes);
        }
        catch (Exception ex)
        {
            error = $"Cannot read the embedded app entry: {ex.Message}";
            return false;
        }

        try
        {
            var actualHash = Convert.ToHexString(
                SHA256.HashData(sourceBytes)).ToLowerInvariant();
            // The digest and the entry it covers ship in the same embedded
            // resource set, so this catches a stale or partial publish rather
            // than tampering; an ordinary comparison is what it calls for.
            if (!string.Equals(expectedHash, actualHash, StringComparison.OrdinalIgnoreCase))
            {
                error = "App entry integrity check failed.";
                return false;
            }
        }
        catch (Exception ex)
        {
            error = $"Cannot verify the embedded app entry: {ex.Message}";
            return false;
        }

        app = new PackagedApp(id, name, version, "main.js", source);
        return true;
    }

    private static int? Int(JsonObject value, string key)
    {
        try { return value[key]?.GetValue<int>(); }
        catch { return null; }
    }

    private static string? Str(JsonObject value, string key)
    {
        try { return value[key]?.GetValue<string>(); }
        catch { return null; }
    }

    private static string Value(JsonObject value, string key) =>
        value[key]?.ToJsonString() ?? "a missing value";
}
